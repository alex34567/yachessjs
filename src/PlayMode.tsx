import * as React from 'react'
import { useEffect, useState } from 'react'
import { Human, Player } from './player'
import { State } from './logic/state'
import { BLACK } from './logic/pieces'
import ChessBoard from './ChessBoard'
import GameInfo from './GameInfo'
import { Pos } from './logic/util'
import { useHistory, useLocation } from 'react-router-dom'
import { changeMode, getStateFromQuery } from './util'
import { ModeProps } from './App'

interface Players {
  white: Player
  black: Player
}

export default function PlayMode (props: ModeProps) {
  const history = useHistory()
  const location = useLocation()
  const [players, setPlayers] = useState<Players>({ white: new Human(), black: new Human() })

  const [state, setState] = useState(() => getStateFromQuery(history))
  const [highlightedPos, setHighlightedPos] = useState<Pos>()
  const [flipBoard, setFlipBoard] = useState(false)

  useEffect(() => {
    // If someone pushes the reset button at the exact time a move is ready,
    // then the board will be corrupted with the move that have been be canceled.
    // This variable blocks the board from updating in the promise handler,
    // preventing this race condition.
    const defaultState = getStateFromQuery(history)
    let stateValid = true
    let player1 = players.white
    let player2 = players.black
    if (defaultState.currTurn === BLACK) {
      const tmp = player1
      player1 = player2
      player2 = tmp
    }
    const makeMoveThen = (state: State): Promise<unknown> | undefined => {
      if (!stateValid) {
        return
      }
      if (state.isGameOver()) {
        setState(state)
        return
      }
      const tmp = player1
      player1 = player2
      player2 = tmp
      const ret = player1.makeMove(state).then(makeMoveThen)
      setState(state)
      return ret
    }
    player1.makeMove(defaultState).then(makeMoveThen)
    setState(defaultState)
    return () => {
      stateValid = false
      players.white.close()
      players.black.close()
    }
  }, [players])

  const restart = (white: Player, black: Player) => {
    setPlayers({ white, black })
  }

  const switchMode = () => {
    changeMode(history, location, state, '/setup.html', false)
  }

  function toggleBoardFlip () {
    setFlipBoard(!flipBoard)
  }

  let currPlayer = players.white
  if (state.currTurn === BLACK) {
    currPlayer = players.black
  }
  return (
    <div className="App">
      <div className="PlayChessBoardBox">
        <ChessBoard flipBoard={flipBoard} changeHighlight={setHighlightedPos} highlightedPos={highlightedPos} makeMove={currPlayer.getBoardClick()} state={state} theme={props.theme}/>
      </div>
      <GameInfo flipBoard={toggleBoardFlip} state={state} restart={restart} switchMode={switchMode} openTheme={props.openTheme}/>
    </div>
  )
}
