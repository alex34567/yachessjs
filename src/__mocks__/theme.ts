import DEFAULT_THEME from '../defaultTheme.json'
import assert from 'assert'
import { Theme } from '../theme'
import { useState } from 'react'

export class ThemeManager {
  readonly theme: Readonly<Theme>
  readonly className: string = 'TestTheme'

  readonly style = document.createElement('style')

  constructor (theme?: Theme) {
    if (!theme) {
      theme = DEFAULT_THEME
    }
    this.theme = { ...theme }
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.whiteColor), 'Attempted xss')
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.blackColor), 'Attempted xss')
    this.style.textContent = `
      .ChessBoardSquareWhite.${this.className} {
        background-color: ${this.theme.boardSquare.whiteColor};
      }

      .ChessBoardSquareBlack.${this.className} {
       background-color: ${this.theme.boardSquare.blackColor};
      }
    `
  }
}

export function useTheme (defaultTheme?: Theme): [ThemeManager, (theme: Theme) => void] {
  const [theme, setTheme] = useState(new ThemeManager(defaultTheme))

  return [theme, theme => {
    setTheme(new ThemeManager(theme))
  }]
}
