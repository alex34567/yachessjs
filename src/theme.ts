import { useEffect, useState } from 'react'
import DEFAULT_THEME from './defaultTheme.json'
import assert from 'assert'

export interface BoardSquareTheme {
  name: string,
  id: string,
  // Must be an css color
  blackColor: string,
  whiteColor: string,
}

export interface PieceTheme {
  prefix: string,
  name: string,
}

export interface Theme {
  piece: PieceTheme
  boardSquare: BoardSquareTheme,
}

let CLASSNAME_SUFFIX = 0

export class ThemeManager {
  readonly theme: Readonly<Theme>
  readonly className: string

  readonly style = document.createElement('style')

  constructor (theme?: Theme) {
    if (!theme) {
      theme = DEFAULT_THEME
    }
    this.theme = { ...theme }
    this.className = `ChessTheme-${CLASSNAME_SUFFIX}`
    CLASSNAME_SUFFIX++
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.whiteColor), 'Attempted xss')
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.blackColor), 'Attempted xss')
    this.style.textContent = `
      .ChessBoardSquareWhite.${this.className} {
        background-color: ${this.theme.boardSquare.whiteColor};
      }

      .ChessBoardSquareBlack.${this.className} {
       background-color: ${this.theme.boardSquare.blackColor};
      }
    `
  }
}

export function useTheme (defaultTheme?: Theme): [ThemeManager, (theme: Theme) => void] {
  const [theme, setTheme] = useState(new ThemeManager(defaultTheme))
  useEffect(() => {
    document.head.append(theme.style)
    return () => {
      theme.style.remove()
    }
  }, [theme])

  return [theme, theme => {
    setTheme(new ThemeManager(theme))
  }]
}
