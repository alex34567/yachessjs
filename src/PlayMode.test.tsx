import { render, unmountComponentAtNode } from 'react-dom'
import { act } from '@testing-library/react'
import React from 'react'
import PlayMode from './PlayMode'
import { ThemeManager } from './theme'
import { BrowserRouter } from 'react-router-dom'

jest.mock('./BoardSquare')
jest.mock('./theme')

let container: HTMLDivElement | null = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container!)
  container!.remove()
  container = null
})

test('Play Mode', () => {
  const openTheme = jest.fn()

  act(() => {
    render(
      <BrowserRouter>
        <PlayMode theme={new ThemeManager()} openTheme={openTheme} />
      </BrowserRouter>
      , container
    )
  })

  expect(container).toMatchSnapshot()
})
