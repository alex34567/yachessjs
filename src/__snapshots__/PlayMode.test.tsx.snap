// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Play Mode 1`] = `
<div>
  <div
    class="App"
  >
    <div
      class="PlayChessBoardBox"
    >
      <div
        class="ChessBoardSizer"
      >
        <div
          class="ChessBoard"
          style="height: 0px; width: 0px;"
        >
          <div
            data-name="r"
          />
          <div
            data-isblack="true"
            data-name="n"
          />
          <div
            data-name="b"
          />
          <div
            data-isblack="true"
            data-name="q"
          />
          <div
            data-name="k"
          />
          <div
            data-isblack="true"
            data-name="b"
          />
          <div
            data-name="n"
          />
          <div
            data-isblack="true"
            data-name="r"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div
            data-isblack="true"
            data-name="p"
          />
          <div
            data-name="p"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-isblack="true"
          />
          <div />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="P"
          />
          <div
            data-isblack="true"
            data-name="R"
          />
          <div
            data-name="N"
          />
          <div
            data-isblack="true"
            data-name="B"
          />
          <div
            data-name="Q"
          />
          <div
            data-isblack="true"
            data-name="K"
          />
          <div
            data-name="B"
          />
          <div
            data-isblack="true"
            data-name="N"
          />
          <div
            data-name="R"
          />
        </div>
      </div>
    </div>
    <div
      class="GameInfo"
    >
      <br />
      <label>
        White Player 
      </label>
      <select>
        <option
          value="human"
        >
          Human
        </option>
        <option
          value="random"
        >
          MrRandom
        </option>
        <option
          value="stockfish"
        >
          Stockfish
        </option>
      </select>
      <br />
      <label>
        Black Player 
      </label>
      <select>
        <option
          value="human"
        >
          Human
        </option>
        <option
          value="random"
        >
          MrRandom
        </option>
        <option
          value="stockfish"
        >
          Stockfish
        </option>
      </select>
      <br />
      <button>
        Restart
      </button>
      <button>
        Setup
      </button>
      <br />
      <button>
        Flip Board
      </button>
      <button>
        Change Theme
      </button>
    </div>
  </div>
</div>
`;
